sw/src/service-worker/service-worker.js

```diff
@@ -0,0 +1,447 @@
+    }
+
+    // Triggered on 'START_RECORDING' message
+    function startRecording(event) {
+        console.debug('[SW] Starting recording')
+        if (!event.data.payload?.sectionId)
+            throw new Error('[SW] No section ID specified to record')
+
+        const clientId = event.source.id // clientId from MessageEvent
+        // Throw error if another recording is in process
+        if (isClientRecording(clientId))
+            throw new Error(
+                "[SW] Can't start a new recording; a recording is already in process"
+            )
+
+        const newClientRecordingState = {
+            sectionId: event.data.payload?.sectionId,
+            pendingRequests: new Map(),
+            fulfilledRequests: new Map(),
+            recordingTimeout: undefined,
+            recordingTimeoutDelay:
+                event.data.payload?.recordingTimeoutDelay || 200,
+            confirmationTimeout: undefined,
+        }
+        clientRecordingStates[clientId] = newClientRecordingState
+
+        // Send confirmation message to client
+        self.clients.get(clientId).then(client => {
+            client.postMessage({ type: swMsgs.recordingStarted })
+        })
+    }
+
+    /** Used to check if a new recording can begin */
+    function isClientRecording(clientId) {
+        return clientId in clientRecordingStates
+    }
+
+    /** Used to check if requests should be handled by recording handler */
+    function isClientRecordingRequests(clientId) {
+        // Don't record requests when waiting for completion confirmation
+        return (
+            isClientRecording(clientId) &&
+            clientRecordingStates[clientId].confirmationTimeout === undefined
+        )
+    }
+
+    /** Request handler during recording mode */
+    function handleRecordedRequest({ request, event }) {
+        const recordingState = clientRecordingStates[event.clientId]
+
+        clearTimeout(recordingState.recordingTimeout)
+        recordingState.pendingRequests.set(request, 'placeholder')
+
+        fetch(request)
+            .then(response => {
+                return handleRecordedResponse(request, response, event.clientId)
+            })
+            .catch(error => {
+                console.error(error)
+                stopRecording(error, event.clientId)
+            })
+    }
+
+    /** Response handler during recording mode */
+    function handleRecordedResponse(request, response, clientId) {
+        const recordingState = clientRecordingStates[clientId]
+        // add response to temp cache - when recording is successful, move to permanent cache
+        const tempCacheKey = getCacheKey('temp', clientId)
+        addToCache(tempCacheKey, request, response)
+
+        // add request to fulfilled. TODO: Handle response for normalized caching
+        // note that request objects can't be stored in IDB (see 'complete recording' function)
+        recordingState.fulfilledRequests.set(request.url, 'placeholder-value')
```

Comment:
This isn't storing the request object though...?

Reply:
True - 'placeholder value' was once the corresponding response object intended to be stored in IndexedDB, but now that just the number of requests is stored in IndexedDB (because the actual response data is stored in Cache Storage), I think this map is unnecessary.
This is a bit of relic that I will refactor 👍